File: gui/static/scheduler_status.js
const byId = (id) => document.getElementById(id);
const lifecycleLabels = { upcoming: "Upcoming", active: "Active", finished: "Finished", empty: "Empty schedule" };
const scheduleDraftState = JSON.parse(byId("schedule-draft-state").textContent);

function formatDateTime(value, options = {}) {
  if (!value) return "—";
  return new Intl.DateTimeFormat(undefined, { dateStyle: "medium", timeStyle: "short", ...options }).format(new Date(value));
}

function relativeTime(value) {
  if (!value) return "";
  const seconds = Math.max(0, Math.round((new Date(value) - new Date()) / 1000));
  if (seconds < 60) return `in ${seconds} sec`;
  if (seconds < 3600) return `in ${Math.ceil(seconds / 60)} min`;
  if (seconds < 86400) return `in ${Math.floor(seconds / 3600)} hr ${Math.ceil((seconds % 3600) / 60)} min`;
  return `in ${Math.ceil(seconds / 86400)} days`;
}

function formatBytes(value) {
  if (value === null || value === undefined) return "—";
  const units = ["B", "KB", "MB", "GB", "TB"];
  let size = value;
  let index = 0;
  while (size >= 1000 && index < units.length - 1) { size /= 1000; index += 1; }
  return `${size.toFixed(index < 3 ? 0 : 1)} ${units[index]}`;
}

function renderHealth(data) {
  const alert = byId("health-alert");
  const showAlert = data.status === "unavailable";
  alert.hidden = !showAlert;
  alert.textContent = showAlert ? data.message : "";
}

function renderSchedule(data) {
  const schedule = data.schedule;
  const warning = byId("schedule-warning");
  warning.hidden = true;
  if (data.schedule_error || data.schedule_is_last_reported || data.status === "invalid_schedule") {
    warning.hidden = false;
    warning.textContent = data.schedule_error ?? (data.schedule_is_last_reported ? "Showing the last reported schedule; live state cannot currently be confirmed." : "The edited schedule was rejected. The scheduler is continuing with the valid schedule shown below.");
  }
  byId("schedule-empty").hidden = Boolean(schedule);
  byId("schedule-dashboard").hidden = !schedule;
  renderScheduleAction(schedule, data.capture_summary);
  if (!schedule) return;

  const lifecycle = byId("schedule-lifecycle");
  lifecycle.className = `lifecycle-badge lifecycle-badge--${schedule.lifecycle}`;
  lifecycle.textContent = lifecycleLabels[schedule.lifecycle] ?? schedule.lifecycle;
  byId("schedule-title").textContent = schedule.run?.name ?? (schedule.lifecycle === "finished" ? "Completed experiment schedule" : "Current experiment schedule");
  byId("schedule-range").textContent = `${schedule.start_date} → ${schedule.end_date} · ${schedule.num_days} day${schedule.num_days === 1 ? "" : "s"}`;
  const runDetails = byId("schedule-run-details");
  const details = [schedule.run?.researcher, schedule.run?.notes].filter(Boolean);
  runDetails.hidden = details.length === 0;
  runDetails.textContent = details.join(" · ");

  const progress = schedule.progress_percent;
  byId("progress-percent").textContent = `${progress.toFixed(1)}%`;
  byId("progress-ring").style.setProperty("--progress", `${progress * 3.6}deg`);
  byId("progress-ring").setAttribute("aria-label", `${progress.toFixed(1)} percent planned schedule progress`);
  byId("progress-count").textContent = `${schedule.elapsed_captures} / ${schedule.total_captures} planned elapsed`;
  byId("progress-day").textContent = schedule.current_day ? `Day ${schedule.current_day} of ${schedule.num_days}` : (schedule.lifecycle === "finished" ? "Schedule complete" : `${schedule.num_days} scheduled days`);
  byId("metric-next").textContent = formatDateTime(schedule.next_capture_at);
  byId("metric-next-relative").textContent = relativeTime(schedule.next_capture_at);
  byId("metric-final").textContent = formatDateTime(schedule.last_capture_at);

  const capture = data.last_capture;
  const captureCard = byId("last-capture-card");
  captureCard.className = capture ? `metric-state metric-state--${capture.status}` : "";
  byId("metric-last-capture").textContent = capture ? capture.status[0].toUpperCase() + capture.status.slice(1) : "No result yet";
  byId("metric-last-capture-time").textContent = capture ? formatDateTime(capture.scheduled_at) : "Waiting for the first capture";
  if (capture?.message) captureCard.title = capture.message;

  renderCaptureResults(data.capture_summary, data.recent_captures ?? []);

  const storage = data.storage;
  const storageCard = byId("storage-card");
  storageCard.className = storage?.used_percent >= 95 ? "metric-state metric-state--critical" : (storage?.used_percent >= 90 ? "metric-state metric-state--warning" : "");
  byId("metric-storage").textContent = storage ? `${formatBytes(storage.free_bytes)} free` : "Unavailable";
  byId("metric-storage-used").textContent = storage ? `${storage.used_percent.toFixed(1)}% used` : "Storage telemetry unavailable";
  const storageMeter = byId("storage-meter");
  storageMeter.hidden = !storage;
  if (storage) {
    const usedPercent = Math.min(100, Math.max(0, storage.used_percent));
    storageMeter.className = `storage-meter${usedPercent >= 95 ? " storage-meter--critical" : (usedPercent >= 90 ? " storage-meter--warning" : "")}`;
    storageMeter.setAttribute("aria-valuenow", usedPercent.toFixed(1));
    storageMeter.setAttribute("aria-valuetext", `${usedPercent.toFixed(1)}% used`);
    byId("storage-meter-fill").style.width = `${usedPercent}%`;
  } else {
    storageMeter.removeAttribute("aria-valuenow");
    storageMeter.removeAttribute("aria-valuetext");
    byId("storage-meter-fill").style.width = "0%";
  }

  const dayStrip = byId("day-strip");
  dayStrip.replaceChildren(...schedule.days.map((day) => {
    const element = document.createElement("div");
    element.className = `experiment-day experiment-day--${day.status}`;
    element.title = `${day.elapsed_captures} of ${day.total_captures} planned captures elapsed`;
    const number = document.createElement("strong"); number.textContent = `Day ${day.number}`;
    const date = document.createElement("span"); date.textContent = new Date(`${day.date}T12:00:00`).toLocaleDateString(undefined, { month: "short", day: "numeric" });
    element.append(number, date); return element;
  }));

  const activity = schedule.daily_activity;
  byId("activity-window").textContent = activity.window_label;
  byId("activity-window-duration").textContent = activity.window_duration_label;
  byId("activity-cadence").textContent = activity.cadence_label;
  byId("activity-peak").textContent = `${activity.peak_captures_per_hour} captures / hour`;
  byId("activity-legend-peak").textContent = `Peak: ${activity.peak_captures_per_hour} captures per hour`;
  const renderHours = (hours) => hours.map((hour) => {
    const tile = document.createElement("div");
    tile.className = `activity-hour${hour.capture_count ? " activity-hour--active" : ""}${hour.intensity_percent >= 45 ? " activity-hour--strong" : ""}`;
    tile.style.setProperty("--activity-intensity", `${0.08 + (hour.intensity_percent / 100) * 0.92}`);
    tile.title = `${hour.label}–${String((hour.hour + 1) % 24).padStart(2, "0")}:00: ${hour.capture_count} planned capture${hour.capture_count === 1 ? "" : "s"} across ${hour.time_point_count} time point${hour.time_point_count === 1 ? "" : "s"}`;
    tile.setAttribute("aria-label", tile.title);
    const label = document.createElement("span"); label.textContent = String(hour.hour).padStart(2, "0");
    const count = document.createElement("strong"); count.textContent = hour.capture_count || "·";
    tile.append(label, count);
    return tile;
  });
  byId("activity-am").replaceChildren(...renderHours(activity.hours.slice(0, 12)));
  byId("activity-pm").replaceChildren(...renderHours(activity.hours.slice(12)));
  byId("daily-summary").textContent = `${schedule.daily_time_points} time points and ${schedule.daily_captures} planned captures per day.`;
  byId("replicate-summary").textContent = `${schedule.replicates} capture${schedule.replicates === 1 ? "" : "s"} per time point, spaced ${schedule.replicate_interval_seconds} seconds apart.`;
  byId("replicate-burst").replaceChildren(...schedule.replicate_offsets.map((replicate) => { const item = document.createElement("div"); const dot = document.createElement("strong"); dot.textContent = replicate.number; const offset = document.createElement("span"); offset.textContent = `+${replicate.offset_seconds}s`; item.append(dot, offset); return item; }));
}

function renderCaptureResults(summary, recent) {
  const section = byId("capture-results");
  section.hidden = !summary;
  if (!summary) return;
  byId("capture-success").textContent = summary.succeeded;
  byId("capture-failed").textContent = summary.failed;
  byId("capture-missed").textContent = summary.missed;
  byId("capture-remaining").textContent = summary.remaining;
  byId("capture-unreported").textContent = summary.elapsed_unreported;
  const shell = byId("recent-captures-shell");
  shell.hidden = recent.length === 0;
  byId("recent-captures").replaceChildren(...recent.map((capture) => {
    const row = document.createElement("div");
    row.className = `recent-capture recent-capture--${capture.status}`;
    const state = document.createElement("strong");
    state.textContent = capture.status[0].toUpperCase() + capture.status.slice(1);
    const time = document.createElement("span");
    time.textContent = formatDateTime(capture.scheduled_at);
    const message = document.createElement("small");
    message.textContent = capture.message;
    row.append(state, time, message);
    return row;
  }));
}

function renderScheduleAction(schedule, captureSummary) {
  const action = byId("schedule-action");
  const emptyLink = byId("schedule-empty-link");
  emptyLink.href = scheduleDraftState === "ready" ? "/schedule/review" : "/schedule";
  emptyLink.textContent = scheduleDraftState === "ready" ? "Review draft" : "Create a schedule";

  action.hidden = true;
  if (!schedule) return;

  let content = null;
  if (scheduleDraftState === "ready") {
    content = ["Schedule draft", "Draft ready for review", "Continue reviewing the saved draft before activating it.", "Review draft", "/schedule/review"];
  } else if (scheduleDraftState === "invalid") {
    content = ["Schedule draft", "Draft needs attention", "The saved draft could not be read. Open the schedule builder to correct it.", "Open schedule builder", "/schedule"];
  } else if (schedule.lifecycle === "finished") {
    const issues = captureSummary && (captureSummary.failed + captureSummary.missed + captureSummary.elapsed_unreported > 0);
    content = issues
      ? ["Experiment complete", "Experiment finished with capture issues", "Review the recorded capture outcomes below before closing this dataset.", "Create next schedule", "/schedule"]
      : ["Experiment complete", "Experiment finished", "This completed schedule remains available below for reference.", "Create next schedule", "/schedule"];
  }
  if (!content) return;

  byId("schedule-action-eyebrow").textContent = content[0];
  byId("schedule-action-title").textContent = content[1];
  byId("schedule-action-copy").textContent = content[2];
  byId("schedule-action-link").textContent = content[3];
  byId("schedule-action-link").href = content[4];
  action.hidden = false;
}

function render(data) {
  renderHealth(data);
  renderSchedule(data);
  document.dispatchEvent(new CustomEvent("scheduler-status-updated", { detail: data }));
}

async function refreshStatus() {
  try {
    const response = await fetch("/api/scheduler/status", { cache: "no-store" });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    render(await response.json());
  } catch (error) {
    console.error(error);
    render({ status: "unavailable", message: "The scheduler status endpoint could not be reached.", last_heartbeat_at: null, age_seconds: null, schedule: null, schedule_error: null, schedule_is_last_reported: false });
  }
}

render(JSON.parse(byId("initial-scheduler-status").textContent));
window.setInterval(refreshStatus, 5000);

File: gui/services/scheduler_status.py
from __future__ import annotations

from collections import Counter
import json
from datetime import date, datetime, timedelta, timezone
from pathlib import Path
from typing import Any
from zoneinfo import ZoneInfo

from scripts.scheduling.heartbeat import HEARTBEAT_STATES
from scripts.scheduling.scheduler import expand_schedule
from scripts.scheduling.run_store import validate_run_metadata


STALE_AFTER = timedelta(seconds=30)


def build_schedule_overview(
    snapshot: dict,
    *,
    now: datetime,
) -> dict[str, Any]:
    """Turn a loaded heartbeat schedule snapshot into dashboard data."""
    tz = ZoneInfo(snapshot["timezone"])
    normalized = {
        "start_date": snapshot["start_date"],
        "num_days": int(snapshot["num_days"]),
        "times": list(snapshot["times"]),
        "replicates": int(snapshot["replicates"]),
        "replicate_interval_seconds": int(
            snapshot["replicate_interval_seconds"]
        ),
    }
    run = validate_run_metadata(snapshot.get("run"))
    schedule_hash = str(snapshot["hash"])
    local_now = now.astimezone(tz)
    run_times = expand_schedule(normalized, tz)
    start_date = date.fromisoformat(normalized["start_date"])
    end_date = start_date + timedelta(days=normalized["num_days"] - 1)

    elapsed = [run_time for run_time in run_times if run_time < local_now]
    remaining = [run_time for run_time in run_times if run_time >= local_now]

    if not run_times:
        lifecycle = "empty"
    elif local_now < run_times[0]:
        lifecycle = "upcoming"
    elif local_now <= run_times[-1]:
        lifecycle = "active"
    else:
        lifecycle = "finished"

    total = len(run_times)
    progress = 0.0 if total == 0 else round(len(elapsed) / total * 100, 1)
    current_day = None
    if start_date <= local_now.date() <= end_date:
        current_day = (local_now.date() - start_date).days + 1

    days = []
    for offset in range(normalized["num_days"]):
        day_date = start_date + timedelta(days=offset)
        day_runs = [value for value in run_times if value.date() == day_date]
        day_elapsed = sum(value < local_now for value in day_runs)
        if day_date < local_now.date():
            day_status = "complete"
        elif day_date == local_now.date():
            day_status = "current"
        else:
            day_status = "upcoming"
        days.append(
            {
                "number": offset + 1,
                "date": day_date.isoformat(),
                "status": day_status,
                "elapsed_captures": day_elapsed,
                "total_captures": len(day_runs),
            }
        )

    daily_activity = build_daily_activity(
        normalized["times"],
        replicates=normalized["replicates"],
    )

    return {
        "lifecycle": lifecycle,
        "hash": schedule_hash,
        "timezone": snapshot["timezone"],
        "start_date": start_date.isoformat(),
        "end_date": end_date.isoformat(),
        "num_days": normalized["num_days"],
        "current_day": current_day,
        "times": normalized["times"],
        "daily_time_points": len(normalized["times"]),
        "replicates": normalized["replicates"],
        "replicate_interval_seconds": normalized[
            "replicate_interval_seconds"
        ],
        "daily_captures": len(normalized["times"]) * normalized["replicates"],
        "total_captures": total,
        "elapsed_captures": len(elapsed),
        "remaining_captures": len(remaining),
        "progress_percent": progress,
        "first_capture_at": _iso_or_none(run_times[0] if run_times else None),
        "next_capture_at": _iso_or_none(remaining[0] if remaining else None),
        "last_capture_at": _iso_or_none(run_times[-1] if run_times else None),
        "days": days,
        "daily_activity": daily_activity,
        "replicate_offsets": [
            {
                "number": index + 1,
                "offset_seconds": index
                * normalized["replicate_interval_seconds"],
            }
            for index in range(normalized["replicates"])
        ],
        "run": run,
    }


def build_daily_activity(
    times: list[str],
    *,
    replicates: int = 1,
) -> dict[str, Any]:
    """Summarize daily capture load as hourly activity and cadence."""
    minutes = sorted(_time_to_minutes(value) for value in times)
    counts = [0] * 24
    for value in minutes:
        counts[value // 60] += 1
    peak_time_points = max(counts, default=0)
    peak_captures = peak_time_points * replicates
    hours = [
        {
            "hour": hour,
            "label": f"{hour:02d}:00",
            "time_point_count": count,
            "capture_count": count * replicates,
            "intensity_percent": (
                0
                if peak_time_points == 0
                else round(count / peak_time_points * 100, 1)
            ),
        }
        for hour, count in enumerate(counts)
    ]

    cadence_kind, cadence_minutes = _summarize_cadence(minutes)
    cadence_label = {
        "empty": "No time points",
        "single": "Single time point",
        "variable": "Variable intervals",
    }.get(cadence_kind)
    if cadence_kind == "regular":
        cadence_label = f"Every {_format_duration(cadence_minutes)}"
    elif cadence_kind == "typical":
        cadence_label = f"Typically every {_format_duration(cadence_minutes)}"

    first = times[0] if times else None
    last = times[-1] if times else None
    window_minutes = minutes[-1] - minutes[0] if minutes else 0
    return {
        "hours": hours,
        "first_time": first,
        "last_time": last,
        "window_minutes": window_minutes,
        "window_duration_label": _format_duration(window_minutes),
        "window_label": (
            f"{first}–{last}" if first is not None and last is not None else "—"
        ),
        "cadence_kind": cadence_kind,
        "cadence_minutes": cadence_minutes,
        "cadence_label": cadence_label,
        "peak_time_points_per_hour": peak_time_points,
        "peak_captures_per_hour": peak_captures,
    }


def _summarize_cadence(minutes: list[int]) -> tuple[str, int | None]:
    if not minutes:
        return "empty", None
    if len(minutes) == 1:
        return "single", None
    deltas = [end - start for start, end in zip(minutes, minutes[1:])]
    cadence, frequency = Counter(deltas).most_common(1)[0]
    if frequency == len(deltas):
        return "regular", cadence
    if frequency / len(deltas) >= 0.75:
        return "typical", cadence
    return "variable", None


def read_scheduler_status(
    heartbeat_path: Path,
    *,
    now: datetime | None = None,
    stale_after: timedelta = STALE_AFTER,
) -> dict[str, Any]:
    """Read scheduler health and its last reported loaded schedule."""
    current_time = now or datetime.now(timezone.utc)
    heartbeat = _read_heartbeat(
        heartbeat_path,
        now=current_time,
        stale_after=stale_after,
    )
    if heartbeat is None:
        return _unavailable_status()
    payload, health = heartbeat

    overview = None
    schedule_error = None
    snapshot = payload.get("schedule")
    if snapshot is not None:
        try:
            overview = build_schedule_overview(snapshot, now=current_time)
        except (KeyError, TypeError, ValueError):
            schedule_error = "The loaded schedule details could not be read."

    return {
        **health,
        "schedule": overview,
        "schedule_error": schedule_error,
        "schedule_is_last_reported": health["status"] == "stale"
        and overview is not None,
        "last_capture": _optional_dict(payload.get("last_capture")),
        "capture_summary": _optional_dict(payload.get("capture_summary")),
        "recent_captures": _dict_list(payload.get("recent_captures")),
        "storage": _optional_dict(payload.get("storage")),
    }


def read_scheduler_health(
    heartbeat_path: Path,
    *,
    now: datetime | None = None,
    stale_after: timedelta = STALE_AFTER,
) -> dict[str, Any]:
    """Read only the scheduler fields needed by the global health pill."""
    heartbeat = _read_heartbeat(
        heartbeat_path,
        now=now or datetime.now(timezone.utc),
        stale_after=stale_after,
    )
    return _unavailable_health() if heartbeat is None else heartbeat[1]


def _read_heartbeat(
    heartbeat_path: Path,
    *,
    now: datetime,
    stale_after: timedelta,
) -> tuple[dict[str, Any], dict[str, Any]] | None:
    try:
        payload = json.loads(heartbeat_path.read_text())
        timestamp = datetime.fromisoformat(payload["timestamp"])
        state = payload["state"]
        message = payload["message"]
        if payload.get("version") != 1 or timestamp.tzinfo is None:
            raise ValueError("unsupported heartbeat")
        if state not in HEARTBEAT_STATES or not isinstance(message, str):
            raise ValueError("invalid heartbeat state")
    except (OSError, ValueError, KeyError, TypeError, json.JSONDecodeError):
        return None

    age_seconds = max(
        0.0,
        (now.astimezone(timezone.utc) - timestamp).total_seconds(),
    )
    if age_seconds > stale_after.total_seconds():
        status = "stale"
        message = "The scheduler heartbeat has stopped updating."
    else:
        status = {
            "running": "healthy",
            "waiting_for_schedule": "waiting_for_schedule",
            "invalid_schedule": "invalid_schedule",
        }[state]

    return payload, {
        "status": status,
        "last_heartbeat_at": timestamp.isoformat(),
        "age_seconds": round(age_seconds, 1),
        "message": message,
    }


def _unavailable_status() -> dict[str, Any]:
    return {
        **_unavailable_health(),
        "schedule": None,
        "schedule_error": None,
        "schedule_is_last_reported": False,
        "last_capture": None,
        "capture_summary": None,
        "recent_captures": [],
        "storage": None,
    }


def _unavailable_health() -> dict[str, Any]:
    return {
        "status": "unavailable",
        "last_heartbeat_at": None,
        "age_seconds": None,
        "message": "No valid scheduler heartbeat is available.",
    }


def _time_to_minutes(value: str) -> int:
    parsed = datetime.strptime(value, "%H:%M")
    return parsed.hour * 60 + parsed.minute


def _format_duration(minutes: int | None) -> str:
    if minutes is None:
        return "—"
    hours, remaining = divmod(minutes, 60)
    if hours and remaining:
        return f"{hours} hr {remaining} min"
    if hours:
        return f"{hours} hr"
    return f"{remaining} min"


def _iso_or_none(value: datetime | None) -> str | None:
    return value.isoformat() if value is not None else None


def _optional_dict(value: Any) -> dict | None:
    return value if isinstance(value, dict) else None


def _dict_list(value: Any) -> list[dict]:
    if not isinstance(value, list):
        return []
    return [item for item in value if isinstance(item, dict)]
